enhancements/update/duration-of-migration-to-multi-arch.md

```diff
@@ -0,0 +1,166 @@
+
+### Non-Goals
+
+Whether or not 'oc adm upgrade' is executed succesfully or the requested upgrade is accepted is not in the scope of this enhancement. So it is _not_ required to have test cases for this enhancement to cover (although they are valid in general):
+- upgrade from multi-arch to single-arch,
+- upgrade from multi-arch to multi-arch,
+- upgrade a cluster in a cloud provider where multi-arch is not supported to multi-arch, or
+- cover the rescue path in case migration to multi-arch hits a failure.
+
+The `oc adm upgrade status` might be useful to monitor the migration progress to multi-arch but the output of the command does not block this enhancement and its implementation. The `status` command is designed for cluster upgrades but migration is not an upgrade conceptually and is implemented through an upgrade.
+
+## Proposal
+
+The proposal includes the steps that an OpenShift cluster operator may take to provide information about the upgrade in additional to the existing names and version numbers for the components that the operator manages.
+
+* Add a few elements into the existing list `ClusterOperator.status.versions` in the `ClusterOperator` manifest. Those new `versions` contains the image pull specifications for the operator and OPTIONALLY the managed operands that the operator chooses to track. Although the version stays the same, the image pull specification are different if a cluster becomes multi-arch.
+
+* Populate the `ClusterOperator.status.versions` including these new elements above.
+
+The CVO has already implemented the way that it checks if each element in the `ClusterVersion.status.versions` from the payload's has the counterpart in the actual one on the cluster.
+
+There is no need to extend the APIs in `openshift/api`, or modify the CVO implementation.
+
+### Workflow Description
+
+This enhancement changes no workflow.
+
+### API Extensions
+
+None.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+[Hosted control planes enables the decoupling of updates between the control plane and the data plane](https://docs.openshift.com/container-platform/4.17/hosted_control_planes/hcp-updating.html#hcp-get-upgrade-versions_hcp-updating):
+
+* Update a control plane: specify `HostedCluster.spec.release`, and
+* Update a data plane: specify `NodePool.spec.release`,
+
+where the value is [the image pull specification of an OCP release payload image](https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1beta1.Release).
+
+The MCO does not run on a hosted cluster and thus our proposal in this enhancement makes no improvement (or regression) for migration of a hosted cluster to multi-arch. Currently there is no way to discover if a hosted cluster is ready to serve a cross-arch node pool. The requests are mainly from the users of standalone clusters at the moment. When hosted cluster users desire the improvement, we can work with the HyperShift team for a solution.
```

Comment:
> Currently there is no way to discover if a hosted cluster is ready to serve a cross-arch node pool.

This is not true. [There is a status condition](https://github.com/openshift/hypershift/blob/e6c368bc9389621cc2d7d4b2d522cc68eaaca4d1/api/hypershift/v1beta1/hostedcluster_types.go#L1258) on the HostedCluster that signals if the HostedCluster can support NodePools of different CPU architectures. (note only one CPU architecture type per NodePool is allowed).

Reply:
My understanding from [this](https://redhat-external.slack.com/archives/C01C8502FMM/p1729529493111299?thread_ts=1729193566.735879&cid=C01C8502FMM) is that "payloadArch" could mean it it or will be ready soon (i.e., still under reconciliation) to serve a secondary arch node pool. And we want a condition that exclude "will be ready".

What did I miss there?

Reply:
Maybe the statement just needs clarified then. As it is currently stated, it seems like one could read "there is no way to tell if a HostedCluster can support NodePools of different CPU architectures" when there is currently a way - checking that status condition and verifying the HC upgrade to a multi-arch release image completed.

It's more that there is no way to determine if a hosted cluster is ready to serve NodePools of varying CPU architectures in the middle of the upgrade process, is that what is trying to be conveyed here?

Reply:
Updated. Let me if I made the correct clarification.

Reply:
@bryan-cox how/when is this payloadArch populated? @hongkailiu when hypershift adds the image version in the status could we use this payloadArch field as a "it is ready" filled all the time for the purposes of CVO?

Reply:
The HyperShift Operator updates the field for the HostedCluster [here](https://github.com/openshift/hypershift/blob/7dd9cb49e3f35408ab451f2c8a1d2f32d62ea681/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L1048).

Reply:
> @hongkailiu when hypershift adds the image version in the status could we use this payloadArch field as a "it is ready" field all the time for the purposes of CVO?

Are you suggesting to use payloadArch=Multi to indicate the hosted cluster is ready for creating a cross arch node pool? We cannot because it could mean it-will-be-ready-soon, i.e., not ready yet. Adding `status` for HyperShift handcrafted COs is not relevant because we are looking for an equivalent signal as `co/machine-config` finished upgrade and it does not exist on hosted clusters.